### PR TITLE
DOC Extending the vocab and storing embeddings

### LINKED ISSUE #1300: Regarding bugs generated by tasks that require the addition of custom tokens.
### System Info

peft==0.7.1

### Who can help?

@BenjaminBossan 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [X] My own task or dataset (give details below)

### Reproduction

This is my raw code, I need to fine-tune the LLAMA model on my dataset, and I need to add some dedicated tokens ，（token_{i}As input, i as label).
```
# LLAMA Tokenizer and model
tokenizer = LlamaTokenizer.from_pretrained(config.checkpoints.path)
model = LlamaForCausalLM.from_pretrained(
    config.checkpoints.path,
    load_in_8bit=True,
    device_map="auto",
    torch_dtype=torch.float16,
)

tokenizer.add_tokens([f"<token_{i}>" for i in range(1025)])

tokenizer.add_tokens([f"{i}" for i in range(257)])

model.resize_token_embeddings(len(tokenizer))

eval_prompt = """
Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.
### Instruction:
task name

### Input:
Front:<token_796><token_301><token_559><token_801><token_272><token_514><token_87><token_272>

### Response:
"""

model_input = tokenizer(eval_prompt, return_tensors="pt").to(model.device)
model.eval()
    with torch.no_grad():
        print("before train")
        print(
            tokenizer.decode(
                model.generate(**model_input, max_new_tokens=100)[0],
                skip_special_tokens=True,
            )
        )
peft_config = LoraConfig(
        task_type=TaskType.CAUSAL_LM,
        inference_mode=False,
        r=8,
        lora_alpha=32,
        lora_dropout=0.05,
        target_modules=["q_proj", "v_proj"],
    )
model = prepare_model_for_kbit_training(model)
model = get_peft_model(model, peft_config)
train_config = {
        "lora_config": peft_config,
        "learning_rate": 1e-4,
        "num_train_epochs": config.train.num_train_epochs,
        "gradient_accumulation_steps": 2,
        "per_device_train_batch_size": 2,
        "gradient_checkpointing": False,
    }
training_args = TrainingArguments(
        output_dir=output_dir,
        # report_to=config.train.report_to,
        overwrite_output_dir=True,
        bf16=True,  # Use BF16 if available
        # logging strategies
        logging_dir=f"{output_dir}/logs",
        logging_strategy="steps",
        logging_steps=10,
        save_strategy="no",
        # optim="adamw_torch_fused",
        optim="paged_adamw_32bit",
        max_steps=total_steps if enable_profiler else -1,
        **{k: v for k, v in train_config.items() if k != "lora_config"},
    )
trainer = Trainer(
            model=model,
            args=training_args,
            train_dataset=train_dataset,
            data_collator=default_data_collator,
            callbacks=[profiler_callback] if enable_profiler else [],
        )

        # Start training
        trainer.train()
model.eval()
    with torch.no_grad():
        print("After train")
        print(
            tokenizer.decode(
                model.generate(**model_input, max_new_tokens=100)[0],
                skip_special_tokens=True,
            )
        )
model.save_pretrained(output_dir)
tokenizer.save_pretrained(output_dir)
```
The above is the code I trained. After the training, I first called generate method to check whether the model was successfully fine-tuned (it could be called normally in the training script). Then I called save_pretrained(output_dir) method to save the model. I found that the save_pretrained(output_dir) method did not hold ["embed_tokens" and "lm_head"] by default. When I tried to use a trained model for inference, I got an cuda error:
```
# This is another file
tokenizer = LlamaTokenizer.from_pretrained("tmp/llama-output")

model = LlamaForCausalLM.from_pretrained(
    "tmp/llama-output",
    load_in_8bit=True,
    device_map="auto",
    torch_dtype=torch.float16,
)
eval_prompt = """
Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.
### Instruction:
task name

### Input:
Front:<token_796><token_301><token_559><token_801><token_272><token_514><token_87><token_272>

### Response:
"""

model_input = tokenizer(eval_prompt, return_tensors="pt").to(model.device)
model.eval()
with torch.no_grad():
    print("After train")
    print(
        tokenizer.decode(
            model.generate(**model_input, max_new_tokens=100)[0],
            skip_special_tokens=True,
        )
    )
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
Loading checkpoint shards: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 6/6 [00:09<00:00,  1.65s/it]
After train
../aten/src/ATen/native/cuda/Indexing.cu:1141: indexSelectLargeIndex: block: [78,0,0], thread: [32,0,0] Assertion `srcIndex < srcSelectDimSize` failed.
../aten/src/ATen/native/cuda/Indexing.cu:1141: indexSelectLargeIndex: block: [78,0,0], thread: [33,0,0] Assertion `srcIndex < srcSelectDimSize` failed.
```
adapter_config.json:
```
{
  "alpha_pattern": {},
  "auto_mapping": null,
  "base_model_name_or_path": "checkpoints/llama2",
  "bias": "none",
  "fan_in_fan_out": false,
  "inference_mode": true,
  "init_lora_weights": true,
  "layers_pattern": null,
  "layers_to_transform": null,
  "loftq_config": {},
  "lora_alpha": 32,
  "lora_dropout": 0.05,
  "megatron_config": null,
  "megatron_core": "megatron.core",
  "modules_to_save": null,
  "peft_type": "LORA",
  "r": 8,
  "rank_pattern": {},
  "revision": null,
  "target_modules": [
    "v_proj",
    "q_proj"
  ],
  "task_type": "CAUSAL_LM"
}
```

### Expected behavior

model.resize_token_embeddings(len(tokenizer)) is usually performed to add custom tokens to the initial LLM. The default save_pretrained() behavior is not saved. I think this is logically missing. The save_pretrained() method should be able to save ["embed_tokens", "lm_head"] by default,That's the first question.

### PR DESCRIPTION
**Pull request by BenjaminBossan**
Resolves #1300

Sourab added the feature to store the embedding layers alongside the adapter in #1147. This PR adds an entry to the documentation to explain the new feature.

### CODE REVIEW

**Review by pacman100 — COMMENTED** [review]
Thank you Benjamin for updating the docs, LGTM! ✨

Btw, one important point during inference is to first load the base model,resize it and then load the PEFT checkpoint. Might be worth mentioning it in docs

**Comment by BenjaminBossan** [comment]
@pacman100 Good point. I added the following sentence:

> For inference, please ensure that you first load the base model, then resize it in the same way as you did before you trained the model, and only then load the PEFT checkpoint.

File: docs/source/developer_guides/troubleshooting.md [hunk 1]
```diff
@@ -81,3 +81,27 @@ You should probably TRAIN this model on a down-stream task to be able to use it
 ```
 
 The mentioned layers should be added to `modules_to_save` in the config to avoid the described problem.
+
+### Extending the vocabulary
+
+For many language fine-tuning tasks, it's necessary to extend the vocabulary of the model, since new tokens are being introduced. When this is done, the embedding layer needs to be extended to account for the new tokens. But this means that the embedding layer should be stored in addition to the adapter weights when saving the adapter.
```
Comment (stevhliu) [inline docs/source/developer_guides/troubleshooting.md:87]:
```suggestion
For many language fine-tuning tasks, extending the model's vocabulary is necessary since new tokens are being introduced. This requires extending the embedding layer to account for the new tokens and also storing the embedding layer in addition to the adapter weights when saving the adapter.
```

File: docs/source/developer_guides/troubleshooting.md [hunk 2]
```diff
@@ -81,3 +81,27 @@ You should probably TRAIN this model on a down-stream task to be able to use it
...
+
+To ensure that this works, add the embedding layer to the `target_modules` of the config. As an example, for Mistral, the config could look like this:
```
Comment (stevhliu) [inline docs/source/developer_guides/troubleshooting.md:89]:
```suggestion
Save the embedding layer by adding it to the `target_modules` of the config. The embedding layer name must follow the standard naming scheme from Transformers. For example, the Mistral config could look like this:
```

File: docs/source/developer_guides/troubleshooting.md [hunk 3]
```diff
@@ -81,3 +81,27 @@ You should probably TRAIN this model on a down-stream task to be able to use it
...
+
+```python
+config = LoraConfig(..., target_modules=["embed_tokens", "lm_head", "q_proj", "v_proj"])
+```
+
+Then, PEFT will automatically take care of storing the embedding layer when saving the adapter. However, for this to work, the name of the embedding layer needs to follow the standard naming scheme of 🤗 Transformers. If it doesn't, you can manually pass `save_embedding_layers=True` when saving the adapter:
```
Comment (stevhliu) [inline docs/source/developer_guides/troubleshooting.md:95]:
```suggestion
Once added to `target_modules`, PEFT automatically stores the embedding layer when saving the adapter if the model has the [`~transformers.PreTrainedModel.get_input_embeddings`] and [`~transformers.PreTrainedModel.get_output_embeddings`]. This is generally the case for Transformers models.

If the model's embedding layer doesn't follow the Transformer's naming scheme, you can still save it by manually passing `save_embedding_layers=True` when saving the adapter:
```

File: docs/source/developer_guides/troubleshooting.md [hunk 4]
```diff
@@ -81,3 +81,27 @@ You should probably TRAIN this model on a down-stream task to be able to use it
...
+
+```python
+model = get_peft_model(...)
+# train the model
+model.save_adapter("my_adapter", save_embedding_layers=True)
+```
+
+This requires the model to have the methods `get_input_embeddings` and `get_output_embeddings`, which is generally the case for 🤗 Transformers models.
```
Comment (stevhliu) [inline docs/source/developer_guides/troubleshooting.md:103]:
```suggestion
```

File: docs/source/developer_guides/troubleshooting.md [hunk 5]
```diff
@@ -81,3 +81,27 @@ You should probably TRAIN this model on a down-stream task to be able to use it
...
+
+For inference, please ensure that you first load the base model, then resize it in the same way as you did before you trained the model, and only then load the PEFT checkpoint.
```
Comment (stevhliu) [inline docs/source/developer_guides/troubleshooting.md:105]:
```suggestion
For inference, load the base model first and resize it the same way you did before you trained the model. After you've resized the base model, you can load the PEFT checkpoint.
```

**Review by stevhliu — APPROVED** [review]
Super cool, thanks for adding! I've made a few suggestions mainly focused on giving users all the info first before running the code to prevent any frustration when copying the code and it doesn't work as expected 🙂

File: docs/source/developer_guides/troubleshooting.md [hunk 6]
```diff
@@ -81,3 +81,28 @@ You should probably TRAIN this model on a down-stream task to be able to use it
...
+
+If the model's embedding layer doesn't follow the Transformer's naming scheme, you can still save it by manually passing `save_embedding_layers=True` when saving the adapter:
+
+```python
+model = get_peft_model(...)
+# train the model
+model.save_adapter("my_adapter", save_embedding_layers=True)
+```
+
+
```
Comment (younesbelkada) [inline docs/source/developer_guides/troubleshooting.md:105]:
```suggestion
```

**Review by younesbelkada — APPROVED** [review]
Thanks for extending the doc with how to correctly use embedding layers!

**Review by pacman100 — APPROVED** [review]
Thank you @BenjaminBossan for iterating, LGTM!
